ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java

```diff
@@ -0,0 +1,231 @@
+ */
+package com.github.ambry.clustermap;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import java.util.List;
+
+
+/**
+ * Metrics for the {@link HelixClusterManager}
+ */
+class HelixClusterManagerMetrics {
+  private final HelixClusterManager.HelixClusterManagerCallback clusterMapCallback;
+  private final MetricRegistry registry;
+
+  public Counter liveInstanceChangeTriggerCount;
+  public Counter externalViewChangeTriggerCount;
+  public Counter instanceConfigChangeTriggerCount;
+  public Counter getWritablePartitionIdsMismatchCount;
+  public Counter hasDatacenterMismatchCount;
+  public Counter getDataNodeIdMismatchCount;
+  public Counter getReplicaIdsMismatchCount;
+  public Counter getDataNodeIdsMismatchCount;
+
+  /**
+   * Metrics for the {@link HelixClusterManager}
+   * @param registry The {@link MetricRegistry} associated with the {@link HelixClusterManager}
+   * @param  clusterMapCallback The {@link HelixClusterManager.HelixClusterManagerCallback} used to query information
+   *                            from {@link HelixClusterManager}
+   */
+  HelixClusterManagerMetrics(MetricRegistry registry,
+      final HelixClusterManager.HelixClusterManagerCallback clusterMapCallback) {
+    this.clusterMapCallback = clusterMapCallback;
+    this.registry = registry;
+    initializeDatacenterMetrics();
+    initializeDataNodeMetrics();
+    initializeDiskMetrics();
+    initializePartitionMetrics();
+    initializeCapacityMetrics();
+    liveInstanceChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "liveInstanceChangeTriggerCount"));
+    externalViewChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "externalViewChangeTriggerCount"));
+    instanceConfigChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "instanceConfigChangeTriggerCount"));
+    getWritablePartitionIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getWritablePartitionsMismatchCount"));
+    hasDatacenterMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "hasDatacenterMismatchCount"));
+    getDataNodeIdMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getDataNodeIdMismatchCount"));
+    getReplicaIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getReplicaIdsMismatchCount"));
+    getDataNodeIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getDataNodeIdsMismatchCount"));
+  }
+
+  /**
+   * Initialize datacenter related metrics.
+   */
+  private void initializeDatacenterMetrics() {
+    Gauge<Long> datacenterCount = new Gauge<Long>() {
+      @Override
+      public Long getValue() {
+        return clusterMapCallback.getDatacenterCount();
```

Comment:
This is carried over from the static cluster manager, where this could just have been initialized once, really. With the dynamic cluster manager, one could introduce a new datacenter dynamically without recycling existing instances (which would happen only rarely, of course). So this, interestingly, has some relevance now.